### PR TITLE
Week 4

### PR DESCRIPTION
**Pull request by EugeneTeu**
Applied algo + ds approach 

- added code to figure out the free slots in a day
- 100% passed test suite

### CODE REVIEW

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,162 @@
...
+    // Start of solution 
+    Collection<String> attendees = request.getAttendees();
+
+    // option for no attendees
+    if (attendees.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    Long durationOfMeeting = request.getDuration();
+    // check if duration too long;
+    if (durationOfMeeting > TimeRange.WHOLE_DAY.duration()) {
```
Comment (fanwgwg) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: As mentioned earlier, you may want to handle to case of 0 :)

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,162 @@
...
+    // we now know where are the free slots in a day
+    Collection<TimeRange> possibleFreeSlots = getFreeSlots(mergedEventList, durationOfMeeting);
+    
+    return possibleFreeSlots;
+  }
+
+  /** 
+   * Filters the attendees events required for this meeting request e.g events that will cause a meeting to fail 
+   */
+  public static Collection<Event> getAttendeesEvent(Collection<String> attendees, Collection<Event> events) {
```
Comment (fanwgwg) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: We may not want these and the below APIs to be public

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,162 @@
...
+    Stream<Event> streamEvents = events.stream();
+    Stream<String> streamAttendees = attendees.stream();
+    // Filter events where the set of attendee whose dont match any of the current attendees
+    streamEvents.filter((event) -> !(streamAttendees.anyMatch((attendeeName) -> event.getAttendees().contains(attendeeName))));
+    Collection<Event> listEvents = streamEvents.collect(Collectors.toList());
+    return listEvents;*/
+    
+    //String[] attendeesArr = attendees.toArray(new String[0]);
+
+    Event[] eventsArr = events.toArray(new Event[0]);
```
Comment (fanwgwg) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:84]: Why does it have to be converted into an array first? It's already a collection. I only see it being iterated when index/order is not important

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,162 @@
...
+      if (currentTimeRange.end() >= nextTimeRange.start()) {
+        int biggerEndTime = Math.max(currentTimeRange.end(), nextTimeRange.end());
+      
+        // we increment the current range e.g [1,4],[2,6] => [1,6]
+        // this might not be the optimum interval e.g ... [2,6], [5,7] ...
+        // we need to continue checking with this currentTimeRange
+        currentTimeRange = TimeRange.fromStartEnd(currentTimeRange.start(), biggerEndTime, false);
+        // if we need to change the time range inside the result list
+        mergedListTimeRange.remove(mergedListTimeRange.size() - 1);
+        mergedListTimeRange.add(currentTimeRange);
```
Comment (fanwgwg) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:137]: I wonder that why we need to remove and add? Can we not have added it in the first place? ArrayList's removal takes O(n) time, making this algo O(n^2), but ideally mergeInterval should be a nlogn algorithm

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,162 @@
...
+
+    return mergedListTimeRange;
+  }
+
+  /**
+   * Gets a list of possible time ranges from the merged events 
+   */
+  public static Collection<TimeRange> getFreeSlots(Collection<TimeRange> mergedEventList, long durationOfMeeting) {
+    List<TimeRange> freeSlotsList = new ArrayList();
+    List<TimeRange> timeRangeList = new ArrayList(mergedEventList);
```
Comment (fanwgwg) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:153]: Why do we need to convert it back to ArrayList? If ArrayList is the datastructure that we want it to be, we can actually make sure that mergeTimeRanges returns ArrayList, and pass in ArrayList here, which would saves some space and time overhead

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (EugeneTeu) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:137]: right good point... i could have done it without adding actually haha

**Pull request merged**